lib/World.luau

```diff
@@ -208,45 +249,92 @@ function World:_updateQueryCache(entityArchetype)
+		storage = storage,
+
+		-- Keep track of all entity ids for fast iteration in queries
+		ownedEntities = {},
+	}
+
+	for index, componentId in components do
+		local associatedArchetypes = world.componentIndex[componentId] or ({} :: any)
+		associatedArchetypes[archetypeId] = index
+		world.componentIndex[componentId] = associatedArchetypes
+
+		componentToStorageIndex[componentId] = index
+		storage[index] = {}
+	end
+
+	world.archetypes[archetypeId] = archetype
+	return archetype
+end
+
+function World._transitionArchetype(
+	self: typeof(World.new()),
+	id: EntityId,
+	myComponents: { [Component]: ComponentInstance }?
+)
 	debug.profilebegin("transitionArchetype")
-	local newArchetype = nil
-	local oldArchetype = self._entityArchetypes[id]
-	local oldStorage
+	if myComponents == nil then
+		-- Remove all components
+		local entityRecord = self.entityIndex[id]
+		if entityRecord == nil then
+			warn("Tried to transitionArchetype a dead entity")
+			return
+		end
 
-	if oldArchetype then
-		oldStorage = self:_getStorageWithEntity(oldArchetype, id)
+		local archetype = entityRecord.archetype
+		if archetype == nil then
+			warn("Already have no archetype")
+			return
+		end
 
-		if not components then
-			oldStorage[oldArchetype][id] = nil
+		for _, storageIndex in archetype.componentToStorageIndex do
+			-- TODO:
+			-- swap remove
+			table.remove(archetype.storage[storageIndex], entityRecord.indexInArchetype)
 		end
-	end
 
-	if components then
-		newArchetype = archetypeOf(unpack(self._entityMetatablesCache[id]))
+		entityRecord.archetype = nil
+		entityRecord.indexInArchetype = nil
 
-		if oldArchetype ~= newArchetype then
-			if oldStorage then
-				oldStorage[oldArchetype][id] = nil
-			end
+		-- TODO:
+		-- This is slow (and unsafe)
+		table.remove(archetype.ownedEntities, table.find(archetype.ownedEntities, id))
+	else
+		local componentIds: { ComponentId } = {}
+		for component in myComponents do
+			table.insert(componentIds, #component)
+		end
 
-			if self._pristineStorage[newArchetype] == nil then
-				self._pristineStorage[newArchetype] = {}
-			end
+		local entityRecord = self.entityIndex[id]
+		if entityRecord == nil then
+			entityRecord = {}
+			self.entityIndex[id] = entityRecord
+		end
+		assert(entityRecord, "Make typechecker happy")
 
-			if self._entityArchetypeCache[newArchetype] == nil then
-				debug.profilebegin("update query cache")
-				self._entityArchetypeCache[newArchetype] = true
-				self:_updateQueryCache(newArchetype)
-				debug.profileend()
-			end
-			self._pristineStorage[newArchetype][id] = components
-		else
-			oldStorage[newArchetype][id] = components
+		-- Find the archetype that matches these components
+		local newArchetypeId = archetypeOf(componentIds)
+		local newArchetype = self.archetypes[newArchetypeId] or createArchetype(self, componentIds)
+
+		-- Add entity to archetype
+		local indexInArchetype = #newArchetype.storage[newArchetype.componentToStorageIndex[componentIds[1]]] + 1
```

Comment:
This is wrong, this is the getting the number of columns in the archetype and adding one. You want the row of the entity, which would be at `ownedEntities`. 
So the index should be `local indexInArchetype = #newArchetype.ownedEntities + 1`